docs/docker.md

```diff
@@ -1,156 +1,101 @@
 Unfamiliar with Docker?  [Start here](https://docs.docker.com/get-started/).
 
-## How-to Guide ##
+## Quick Start (testing installation) ##
 
-### Install CDash
+The following instructions spin up CDash & MySQL for local experimentation.
 
-#### Quick Start
-If you haven't done so already, begin by cloning the CDash repository. Then start the 
-`docker-compose` process to accept default values and start using the website
+1. If you haven't done so already, begin by cloning the CDash repository:
 
 ```bash
 git clone https://github.com/Kitware/CDash
 cd CDash
-docker-compose pull
-docker-compose up -d cdash
 ```
 
-#### Customizing the CDash instance
-In the root of your CDash clone, edit `docker-compose.yml`.
+2. Next, use the `docker compose up` command to spin up your new CDash instance:
 
-The `CDASH_CONFIG` section is where you specify settings that will be stored in your `config.local.php`
-file.
-
-Once you're happy with your changes to this file, run:
-
-```
-docker-compose up -d cdash
+```bash
+docker compose -f docker/docker-compose.yml \
+               -f docker/docker-compose.dev.yml \
+               -f docker/docker-compose.mysql.yml \
+               --env-file .env.dev up -d
 ```
 
-This tells [Docker Compose](https://docs.docker.com/compose/) to build and run services for the CDash web server and its MySQL database. This command downloads [a prebuilt image from DockerHub](https://hub.docker.com/r/kitware/cdash/).  If you prefer to build your own Docker image for CDash, pass the `--build` option to `docker-compose`.
+3. Browse to http://localhost:8080.  You should see a freshly installed copy of CDash with the latest database schema.
 
-This initial `docker-compose` command does not run the CDash's install script by default.  To achieve that, run:
-
-```
-docker-compose run --rm cdash install configure
+#### Running the test suite
+```bash
+docker exec -it cdash /bin/bash
+cd _build
+ctest
 ```
 
-This executes a one-shot container that runs the install procedure and sets up the predefined users from your `docker-compose.yml` file.
+## Configuration
 
-Once this command complete, browse to localhost:8080.  You should see a freshly installed copy of CDash with the latest database schema.
+### Why so many YAML files?
+You may have noticed that CDash's `docker compose` configuration is [split across multiple files](https://docs.docker.com/compose/extends/). The allows us to support various workflows (MySQL vs. Postgres, production vs. development) while minimizing code duplication.
 
-#### Change the config and redeploy
+For example, to use Postgres instead of MySQL, pass `-f docker/docker-compose.postgres.yml` instead of `-f docker/docker-compose.mysql.yml` to the `docker compose` commands mentioned in this document.
 
-Edit `docker-compose.yml` and run
+### Changing the default configuration
+You can change the following environment variables in `docker/docker-compose.yml`:
```

Comment:
Do we want to recommend that people edit the docker compose files manually, or should we recommend that they edit their `.env` file?

Reply:
This isn't a bad idea, but it kinda opens up a can of worms. We need to carefully distinguish between the  `.env` that should be used in the build context vs. the `.env` file that should be used for the running container.

Reply:
Unless I'm mistaken, they are essentially be the same.  My understanding is that docker compose reads environment variables from a `.env` or from the environment field in the `.yml` file, and then passes those into the container, so there should never be a need to edit a `.env` in a running container.

Reply:
Your understanding is correct. Nonetheless, I'm running into problems using a single `.env` for both CDash and docker compose.

More specifically, Laravel expects the explicit line `APP_KEY=` (with no value) to be in its `.env` file before you run `artisan key:generate` to generate an application key. If you have this line in your host `.env` file, then the container ends up with an env var `APP_KEY` with no value, and this overrides the `APP_KEY` set in the container's `.env` file. The result is HTTP 500: No application encryption key has been specified.

In your opinion, which solution is less bad?
1. Tell the user to edit `docker-compose.yml` to change the default admin credentials
2. Use separate .env files for docker compose & CDash. CDash's `.env` would remain more-or-less the same, docker compose's file would have env vars like `DB_PASSWORD` that are used in our other docker compose YAML files.

I'm open to other ideas as well, of couse.

Reply:
I agree that having two different `.env` files is confusing and is probably worse than editing `docker-compose.yml`.  If the `APP_KEY` stuff is the only think blocking having a single `.env`, I think we can probably find a workaround eventually.  I don't think it should be a blocker for getting this merged though--we can perhaps recommend option 1 listed above for this release and smooth out some of the rough edges in the future.